In [ ]:
import pandas as pd
import numpy as np

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
mkt = pd.read_excel('/content/drive/MyDrive/WeGrowTest/Database - Q3_2020.xlsx', sheet_name = 'MKT')
sales = pd.read_excel('/content/drive/MyDrive/WeGrowTest/Database - Q3_2020.xlsx', sheet_name = 'Sales')
vd = pd.read_excel('/content/drive/MyDrive/WeGrowTest/Database - Q3_2020.xlsx', sheet_name = 'Vận đơn')

# **Data cleaning**

Dataset contains information about the operations of a Vietnamese company in the third quarter of 2020, including 3 tables:
*   MKT: Contains information about Marketing activities
*   Sales: Contains information about sales activities
*   Vận đơn: Contains information about bill of lading activities







## **1- MKT table**

In [ ]:
#Copy
mkt_1 = mkt.copy()
sales_1 = sales.copy()
vd_1 = vd.copy()

### **Data infomation**

In [ ]:
#Shape
num_rows, num_cols = mkt_1.shape
print('Num rows: ',num_rows)
print('Num columns: ',num_cols)

Num rows:  1511
Num columns:  21


In [ ]:
pd.set_option('display.max_columns', None)
mkt_1.head()

Date Channel    MKTer Chiến dịch  Chi phí Marketing  Impression  \
0 2020-07-01      FB    DucVu        CVS                  0           0   
1 2020-07-01      FB    DucVu       Mess                  0           0   
2 2020-07-01      FB    DucVu      Brand             259352       61673   
3 2020-07-01      FB  NgocNgo        CVS            4541746       95958   
4 2020-07-01      FB  NgocNgo      Brand                  0           0   

     Reach  Click  Share  Cmt Inbox  Lead MKT  Đơn hàng     Doanh thu  \
0      0.0      0      0    0     0         1   1.00000  4.990000e+05   
1      0.0      0      0    0     0         3   1.00000  4.990000e+05   
2  59968.0    271      0    0     0         0   0.00000  0.000000e+00   
3  76992.0    386      4   26     0        40  31.73913  1.583783e+07   
4      0.0      0      0    0     0         0   0.00000  0.000000e+00   

   Paid Revenue 1   Giá Lead  Đơn/Lead           CPM           CPC  \
0          499000       0.00  1.000000      0.000000      0.000000   
1          499000       0.00  0.333333      0.000000      0.000000   
2               0       0.00  0.000000   4205.276215    957.018450   
3        12974000  113543.65  0.793478  47330.561287  11766.181347   
4               0       0.00  0.000000      0.000000      0.000000   

   Giá Mess\n(Cmt + Inbox)  ME1 (ME đã bao gồm thuế + phí)  
0                 0.000000                            0.00  
1                 0.000000                            0.00  
2                 0.000000                       280100.16  
3            174682.538462                      4905085.68  
4                 0.000000                            0.00

This data has 21 columns, including:


*   Date:The date the marketing campaign was implemented
*   Channel: Channel used to display ads
*   MKTer: Name of staff member
*   Chiến dịch: Name of marketing campaign
*   Chi phí Marketing: Marketing expense
*   Impression: Number of ad impressions
*   Reach: Total number of unique individuals exposed to the ad
*   Click: Number of clicks on ad
*   Share: Number of shares
*   Cmt: Number of comments
*   Inbox: Number of inboxs
*   Lead MKT: Number of Leads
*   Đơn hàng: Numbers of orders
*   Doanh thu: Revenue
*   Paid Revenue 1: revenue generated from products that were sold through marketing campaign
*   Giá/Lead: Expense per Lead
*   Đơn/Lead: Order per Lead
*   CPM: Expense per 1000 Impressions
*   CPC: Expense per Click
*   Giá Mess\n(Cmt + Inbox): Expense per Cmt or Inbox
*   ME1: Total marketing expenses and taxes








In [ ]:
mkt_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1511 entries, 0 to 1510
Data columns (total 21 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   Date                            1511 non-null   datetime64[ns]
 1   Channel                         1511 non-null   object        
 2   MKTer                           1511 non-null   object        
 3   Chiến dịch                      1511 non-null   object        
 4   Chi phí Marketing               1511 non-null   int64         
 5   Impression                      1511 non-null   int64         
 6   Reach                           1511 non-null   float64       
 7   Click                           1511 non-null   int64         
 8   Share                           1511 non-null   int64         
 9   Cmt                             1511 non-null   int64         
 10  Inbox                           1511 non-null   object        
 11  Lead





*   Column 'ME1' contains null values
*   Some columns have incorrect dtype





### **Handle null values**

In [ ]:
mkt_1.isna().sum()

Date                               0
Channel                            0
MKTer                              0
Chiến dịch                         0
Chi phí Marketing                  0
Impression                         0
Reach                              0
Click                              0
Share                              0
Cmt                                0
Inbox                              0
Lead MKT                           0
Đơn hàng                           0
Doanh thu                          0
Paid Revenue 1                     0
Giá Lead                           0
Đơn/Lead                           0
CPM                                0
CPC                                0
Giá Mess\n(Cmt + Inbox)            0
ME1 (ME đã bao gồm thuế + phí)    14
dtype: int64

Column ME1 contains 14 null values. This portion of data accounts for a small percentage, and there is no tax information so it is impossible to calculate, so I decided to delete these null values.

In [ ]:
#Drop Null Values
mkt_1 = mkt_1.dropna(axis = 1)

### **Check Duplicate**





In [ ]:
#Check duplicate
duplicated_data = mkt_1[mkt_1.duplicated()]
len(duplicated_data)

1

In [ ]:
duplicated_data = mkt_1[mkt_1.duplicated()]
duplicated_data

Date Channel  MKTer Chiến dịch  Chi phí Marketing  Impression  \
636 2020-08-05      FB  DucVu       Mess                  0           0   

     Reach  Click  Share  Cmt Inbox  Lead MKT  Đơn hàng  Doanh thu  \
636    0.0      0      0    0     0         0       0.0        0.0   

     Paid Revenue 1  Giá Lead  Đơn/Lead  CPM  CPC  Giá Mess\n(Cmt + Inbox)  
636               0       0.0       0.0  0.0  0.0                      0.0

This is not duplicate data, so we don't need to drop it

### **Change Dtype**

In [ ]:
#Column 'Đơn hàng': to int
mkt_1['Đơn hàng'] = round(mkt_1['Đơn hàng'], 0)
mkt_1['Đơn hàng'] = mkt_1['Đơn hàng'].astype(int)
#Column 'Đơn/Lead': to int
mkt_1['Đơn/Lead'] = round(mkt_1['Đơn/Lead'], 0)
mkt_1['Đơn/Lead'] = mkt_1['Đơn/Lead'].astype(int)
#Column 'Inbox': to int
mkt_1['Inbox'] = mkt_1['Inbox'].astype(int)

The 'inbox' column can't be converted to int, so I need to check the values in this column

In [ ]:
#Unique Values
mkt_1['Inbox'].unique()

array(['0', '1', '18', '-'], dtype=object)

In [ ]:
mkt_1[mkt_1['Inbox'] == '-']

Date Channel      MKTer Chiến dịch  Chi phí Marketing  Impression  \
1260  2020-09-14      FB  GiangPhan        CVS            2021614       31348   

      Reach  Click  Share  Cmt Inbox  Lead MKT  Đơn hàng     Doanh thu  \
1260  28080    130      5    9     -        15        15  7.463304e+06   

      Paid Revenue 1       Giá Lead  Đơn/Lead           CPM           CPC  \
1260         6986000  134774.266667         1  64489.409213  15550.876923   

      Giá Mess\n(Cmt + Inbox)  
1260                      0.0

Data has a record with value '-' in column 'Inbox'. Drop this record.

In [ ]:
#Drop Record
mkt_1 = mkt_1.drop(mkt_1[mkt_1['Inbox'] == '-'].index)

In [ ]:
#Change Dtype 'Inbox'
mkt_1['Inbox'] = mkt_1['Inbox'].astype(int)

In [ ]:
#Save file
mkt_1.to_excel('MKT_cleaned.xlsx', index = False)


## **2- Sales table**

In [ ]:
#Copy
sales_1 = sales.copy()

### **Data information**

In [ ]:
num_rows, num_cols = sales_1.shape
print('Num rows: ',num_rows)
print('Num cols: ',num_cols)

Num rows:  9675
Num cols:  21


In [ ]:
sales_1.head()

Unnamed: 0                 Giờ        Khách hàng          SĐT Channel  \
0   1-7-2020 2023-07-01 00:13:42     Pham thu thuy  123400000.0      FB   
1   1-7-2020 2023-07-01 00:17:53  Hoàng Thị Lương   123400001.0      FB   
2   1-7-2020 2023-07-01 02:16:56    Phạm Thị huong  123400002.0      FB   
3   1-7-2020 2023-07-01 05:21:32               Lan  123400003.0      FB   
4   1-7-2020 2023-07-01 05:21:31       Lại Thị Đua  123400004.0      FB   

  Chiến dịch     Content   Marketer 2 Type of Lead  \
0        CVS  B1-1812-P1  AnhNguyen08      Dathang   
1        CVS         cay      NgocNgo      Dathang   
2        CVS      matcon      NgocNgo      Dathang   
3        CVS         cay      NgocNgo      Dathang   
4        CVS  condau_sto  YenNguyen02      Dathang   

  Sales Admin xác nhận Type of Lead     Sales  Số lần tương tác   Ngày gọi  \
0                          DATTHANG  ThuyTran               1.0 2020-07-01   
1                          DATTHANG  ThuyTran               1.0 2020-07-01   
2                          DATTHANG  ThuyTran               1.0 2020-07-03   
3                          DATTHANG  ThuyTran               1.0 2020-07-03   
4                          DATTHANG  ThuyTran               1.0 2020-07-03   

             Trạng thái                               Level Ngày hẹn gọi lại  \
0  Hủy đơn sau khi chốt         L7.2 - Hủy đơn sau khi chốt              NaT   
1              Đặt hàng  L8.2 - Khách đã thanh toán toàn bộ              NaT   
2  Hủy đơn sau khi chốt         L7.2 - Hủy đơn sau khi chốt              NaT   
3  Hủy đơn sau khi chốt         L7.2 - Hủy đơn sau khi chốt              NaT   
4              Đặt hàng  L8.2 - Khách đã thanh toán toàn bộ              NaT   

  Close Date     Tỉnh/TP  Số lượng bộ sách  Số tiền giảm giá  Tổng tiền  
0        NaT         NaN               1.0          140000.0     499000  
1 2023-07-04  Quảng Ninh               1.0          140000.0     499000  
2        NaT         NaN               1.0          140000.0     499000  
3        NaT         NaN               1.0          140000.0     499000  
4 2023-07-05      Hà Nội               1.0          140000.0     499000

Data has 21 columns, including:


*   'Unnamed: 0' : Unclear
*   Giờ: The time that customers leave their information on the ads
*   Khách hàng: Name of Lead
*   SĐT: Number phone of Lead
*   Channel: Channel which customer is advertised
*   Chiến dịch: Marketing campaign which customer is advertised
*   Content: Unclear
*   Marketer 2: Name of marketer
*   Type of Lead: Type of Lead, includes 2 values: Dathang(ordered) and Tuvan (being consulted)
*   Sales Admin xác nhận Type of Lead: Confirmation of type of Lead by Sales Admin
*   Sales: Name of Sales
*   Số lần tương tác: Number of calls
*   Ngày gọi: The day the employee contacted the Lead
*   Trạng thái: Lead status
*   Level: Lead status in more detail
*   Ngày hẹn gọi lại: The date the customer makes an appointment to call back
*   Close date: The date the customer completes payment
*   Tỉnh/TP: Province/City where the customer lives
*   Số lượng bộ sách: Number of book sets ordered
*   Số tiền giảm giá: Discount amount
*   Tổng tiền: Total amount that the customer needs to pay





In [ ]:
sales_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9675 entries, 0 to 9674
Data columns (total 21 columns):
 #   Column                             Non-Null Count  Dtype         
---  ------                             --------------  -----         
 0   Unnamed: 0                         9504 non-null   object        
 1   Giờ                                9504 non-null   datetime64[ns]
 2   Khách hàng                         9504 non-null   object        
 3   SĐT                                9504 non-null   float64       
 4   Channel                            9504 non-null   object        
 5   Chiến dịch                         9504 non-null   object        
 6   Content                            9500 non-null   object        
 7   Marketer 2                         6781 non-null   object        
 8   Type of Lead                       9498 non-null   object        
 9   Sales Admin xác nhận Type of Lead  5448 non-null   object        
 10  Sales                              9

In [ ]:
#Null Values
sales_1.isna().sum()

Unnamed: 0                            171
Giờ                                   171
Khách hàng                            171
SĐT                                   171
Channel                               171
Chiến dịch                            171
Content                               175
Marketer 2                           2894
Type of Lead                          177
Sales Admin xác nhận Type of Lead    4227
Sales                                 171
Số lần tương tác                      179
Ngày gọi                             9471
Trạng thái                            357
Level                                1691
Ngày hẹn gọi lại                     9627
Close Date                           3514
Tỉnh/TP                              6281
Số lượng bộ sách                     2624
Số tiền giảm giá                     2633
Tổng tiền                               0
dtype: int64

**Through observing the data, we noticed the following problems:**


*   The first column has no name and wrong dtype
*   Column 'Giờ' has date data and has a year part as 2023 instead of 2020
*   Some columns has wrong dtype
*   Column 'Closed Date' has a year part as 2023 instead of 2020
*   This data has many columns containing null values








### **Adjust the information and data in the columns**

***Column 'Unnamed : 0'*** has date data and it is similar to the datepart in the column 'Giờ', so I think this column contains information about the date that customers leave their information on ads.

*Change name and dtype for column 'Unnamed : 0'*





In [ ]:
#Dtype
sales_1['Unnamed: 0'] = pd.to_datetime(sales_1['Unnamed: 0'], dayfirst = True)
#Rename
sales_1 = sales_1.rename(columns = {'Unnamed: 0': 'Ngày'})

***Column 'Giờ'***:  *delete date part*





In [ ]:
sales_1['Giờ'] = sales_1['Giờ'].dt.time

***Column 'Close Date'***: *change the year part to 2020*

In [ ]:
#Check Values
sales_1['Năm'] = sales_1['Close Date'].dt.year
sales_1['Năm'].value_counts()

Năm
2023.0    6161
Name: count, dtype: int64

All year parts in this column are 2023

In [ ]:
#Drop sales['Năm]
del sales_1['Năm']
#Change Values
sales_1['Close Date'] = sales_1['Close Date'].apply(lambda x: x.replace(year=2020))

### **Check duplicate**

In [ ]:
data_duplicate = sales_1[sales_1.duplicated()]
data_duplicate

Ngày  Giờ Khách hàng  SĐT Channel Chiến dịch Content Marketer 2  \
9505  NaT  NaN        NaN  NaN     NaN        NaN     NaN        NaN   
9506  NaT  NaN        NaN  NaN     NaN        NaN     NaN        NaN   
9507  NaT  NaN        NaN  NaN     NaN        NaN     NaN        NaN   
9508  NaT  NaN        NaN  NaN     NaN        NaN     NaN        NaN   
9509  NaT  NaN        NaN  NaN     NaN        NaN     NaN        NaN   
...   ...  ...        ...  ...     ...        ...     ...        ...   
9670  NaT  NaN        NaN  NaN     NaN        NaN     NaN        NaN   
9671  NaT  NaN        NaN  NaN     NaN        NaN     NaN        NaN   
9672  NaT  NaN        NaN  NaN     NaN        NaN     NaN        NaN   
9673  NaT  NaN        NaN  NaN     NaN        NaN     NaN        NaN   
9674  NaT  NaN        NaN  NaN     NaN        NaN     NaN        NaN   

     Type of Lead Sales Admin xác nhận Type of Lead Sales  Số lần tương tác  \
9505          NaN                               NaN   NaN               NaN   
9506          NaN                               NaN   NaN               NaN   
9507          NaN                               NaN   NaN               NaN   
9508          NaN                               NaN   NaN               NaN   
9509          NaN                               NaN   NaN               NaN   
...           ...                               ...   ...               ...   
9670          NaN                               NaN   NaN               NaN   
9671          NaN                               NaN   NaN               NaN   
9672          NaN                               NaN   NaN               NaN   
9673          NaN                               NaN   NaN               NaN   
9674          NaN                               NaN   NaN               NaN   

     Ngày gọi Trạng thái Level Ngày hẹn gọi lại Close Date Tỉnh/TP  \
9505      NaT        NaN   NaN              NaT        NaT     NaN   
9506      NaT        NaN   NaN              NaT        NaT     NaN   
9507      NaT        NaN   NaN              NaT        NaT     NaN   
9508      NaT        NaN   NaN              NaT        NaT     NaN   
9509      NaT        NaN   NaN              NaT        NaT     NaN   
...       ...        ...   ...              ...        ...     ...   
9670      NaT        NaN   NaN              NaT        NaT     NaN   
9671      NaT        NaN   NaN              NaT        NaT     NaN   
9672      NaT        NaN   NaN              NaT        NaT     NaN   
9673      NaT        NaN   NaN              NaT        NaT     NaN   
9674      NaT        NaN   NaN              NaT        NaT     NaN   

      Số lượng bộ sách  Số tiền giảm giá  Tổng tiền  
9505               NaN               NaN          0  
9506               NaN               NaN          0  
9507               NaN               NaN          0  
9508               NaN               NaN          0  
9509               NaN               NaN          0  
...                ...               ...        ...  
9670               NaN               NaN          0  
9671               NaN               NaN          0  
9672               NaN               NaN          0  
9673               NaN               NaN          0  
9674               NaN               NaN          0  

[170 rows x 21 columns]



This dataset has 170 duplicated columns

In [ ]:
#Drop Duplicate
sales_1 = sales_1.drop_duplicates()

### **Handle null values**

In [ ]:
#CheckNull
sales_1.isna().sum()

Ngày                                    1
Giờ                                     1
Khách hàng                              1
SĐT                                     1
Channel                                 1
Chiến dịch                              1
Content                                 5
Marketer 2                           2724
Type of Lead                            7
Sales Admin xác nhận Type of Lead    4057
Sales                                   1
Số lần tương tác                        9
Ngày gọi                             9301
Trạng thái                            187
Level                                1521
Ngày hẹn gọi lại                     9457
Close Date                           3344
Tỉnh/TP                              6111
Số lượng bộ sách                     2454
Số tiền giảm giá                     2463
Tổng tiền                               0
dtype: int64

***Drop 3 columns:***

*   'Sales Admin xác nhận Type of Lead': contain the same value as column 'Type of Lead'
*   'Ngày gọi': The data in this column is almost all null
*   'Ngày hẹn gọi lại': The data in this column is almost all null





In [ ]:
#Drop Columns
sales_1 = sales_1.drop(columns=['Sales Admin xác nhận Type of Lead','Ngày gọi','Ngày hẹn gọi lại'])

#### **Check each column that contain null values**









##### **1- Column 'Ngày'**

In [ ]:
df = sales_1[sales_1['Ngày'].isna()]
df

Ngày  Giờ Khách hàng  SĐT Channel Chiến dịch Content Marketer 2  \
9504  NaT  NaT        NaN  NaN     NaN        NaN     NaN        NaN   

     Type of Lead Sales  Số lần tương tác Trạng thái Level Close Date Tỉnh/TP  \
9504          NaN   NaN               NaN        NaN   NaN        NaT     NaN   

      Số lượng bộ sách  Số tiền giảm giá  Tổng tiền  
9504               NaN               NaN          0

The data in this row is all null values. So drop this row.

In [ ]:
sales_1 = sales_1.drop(sales_1[sales_1['Ngày'].isna()].index)



Columns 'Giờ','Khách hàng','SĐT','Channel','Chiến dịch','Sales' now has no null values.



##### **2- Column Content**



In [ ]:
sales_1[sales_1['Content'].isna()]

Ngày       Giờ  Khách hàng          SĐT Channel Chiến dịch Content  \
4228 2020-09-23  10:43:08       Hường  123404228.0      FB        CVS     NaN   
6004 2020-07-29  12:20:17   Nguyệt Vũ  123406002.0      FB        CVS     NaN   
8175 2020-06-08  13:39:25  Thùy Khiêm  123408173.0      FB        CVS     NaN   
9114 2020-09-07  06:22:28     Bui Nga  123409112.0      FB        CVS     NaN   

       Marketer 2 Type of Lead Sales Admin xác nhận Type of Lead  ...  \
4228          NaN      Dathang                               NaN  ...   
6004  YenNguyen02      Dathang                          DATTHANG  ...   
8175  YenNguyen02        Tuvan                               NaN  ...   
9114    GiangPhan      Dathang                          DATTHANG  ...   

     Số lần tương tác  Ngày gọi Trạng thái  \
4228              2.0       NaT   Đặt hàng   
6004              1.0       NaT   Đặt hàng   
8175              5.0       NaT   Đặt hàng   
9114              1.0       NaT   Đặt hàng   

                                   Level Ngày hẹn gọi lại Close Date  \
4228  L8.2 - Khách đã thanh toán toàn bộ              NaT 2020-09-27   
6004  L8.2 - Khách đã thanh toán toàn bộ              NaT 2020-08-01   
8175             L7.1 - Chờ chuyển khoản              NaT        NaT   
9114                                 NaN              NaT 2020-07-11   

         Tỉnh/TP Số lượng bộ sách  Số tiền giảm giá  Tổng tiền  
4228         NaN              1.0          140000.0     499000  
6004    Bắc Ninh              1.0          140000.0     499000  
8175  Bình Dương              1.0          240000.0     399000  
9114      Hà Nội              1.0          140000.0     499000  

[4 rows x 21 columns]

In [ ]:
#ValueCounts
sales_1['Content'].value_counts()

lop11            2577
digheconchong     524
condau_sto        503
ac                423
haoi              394
                 ... 
tonthuong           1
100baihocanh1       1
vietkieu_an         1
chidau              1
vidkol              1
Name: Content, Length: 353, dtype: int64

The column has a lot of distinct values and has no pattern, so I drop null values

In [ ]:
#Drop
sales_1 = sales_1.drop(sales_1[sales_1['Content'].isna()].index)

##### **3- Column 'Marketer 2'**

In [ ]:
sales_1[sales_1['Marketer 2'].isna()]

Ngày       Giờ              Khách hàng          SĐT Channel  \
488  2020-09-10  00:40:53      Nguyen thanh huong  123400488.0      FB   
489  2020-09-10  01:03:24             Nguyễn Lily  123400489.0      FB   
490  2020-09-10  01:27:54  Nguyen Thi Hương Diễm   123400490.0      FB   
491  2020-09-10  02:17:04       Thuylien Thuylien  123400491.0      FB   
492  2020-09-10  04:21:40     Nguyễn Đặng Thanh Ý  123400492.0      FB   
...         ...       ...                     ...          ...     ...   
9499 2020-09-09  09:28:02         Lê thị thu hiền  123409497.0  Google   
9500 2020-09-09  15:30:20         Nguyễn Tiết Hài  123409498.0  Google   
9501 2020-09-09  14:49:21          vương thị hợp   123409499.0  Google   
9502 2020-09-09  18:44:41            Cao thị hạnh  123409500.0  Google   
9503 2020-09-09  20:28:09            Lô Thị Nhung  123409501.0  Google   

     Chiến dịch              Content Marketer 2 Type of Lead     Sales  \
488         CVS                lop11        NaN      Dathang  ThuyTran   
489         CVS                lop11        NaN      Dathang  ThuyTran   
490         CVS               utttkc        NaN      Dathang  ThuyTran   
491         CVS         bosachxizach        NaN      Dathang  ThuyTran   
492         CVS                lop11        NaN      Dathang  ThuyTran   
...         ...                  ...        ...          ...       ...   
9499     PerMax      roasthap_151022        NaN      Dathang  ThuyTran   
9500     PerMax                  909        NaN      Dathang  ThuyTran   
9501     Search                 2809        NaN      Dathang  ThuyTran   
9502     Search  2409-04-01 00:00:00        NaN      Dathang  ThuyTran   
9503     Search                 2809        NaN        Tuvan   NamPham   

      Số lần tương tác            Trạng thái  \
488                1.0              Đặt hàng   
489                1.0              Đặt hàng   
490                1.0  Hủy đơn sau khi chốt   
491                1.0              Đặt hàng   
492                1.0              Đặt hàng   
...                ...                   ...   
9499               1.0              Đặt hàng   
9500               1.0              Đặt hàng   
9501               1.0              Đặt hàng   
9502               1.0              Đặt hàng   
9503               1.0                Sai số   

                                   Level Close Date Tỉnh/TP  Số lượng bộ sách  \
488                       L5.2 - Từ chối        NaT     NaN               1.0   
489                       L5.2 - Từ chối        NaT     NaN               1.0   
490          L7.2 - Hủy đơn sau khi chốt        NaT     NaN               1.0   
491              L7.1 - Chờ chuyển khoản        NaT     NaN               1.0   
492   L8.2 - Khách đã thanh toán toàn bộ 2020-10-02     NaN               1.0   
...                                  ...        ...     ...               ...   
9499  L8.2 - Khách đã thanh toán toàn bộ 2020-09-13     NaN               1.0   
9500                      L5.2 - Từ chối        NaT     NaN               1.0   
9501                      L5.2 - Từ chối        NaT     NaN               1.0   
9502  L8.2 - Khách đã thanh toán toàn bộ 2020-10-03     NaN               1.0   
9503                                 NaN        NaT     NaN               NaN   

      Số tiền giảm giá  Tổng tiền  
488           140000.0     499000  
489           140000.0     499000  
490           140000.0     499000  
491           140000.0     499000  
492           140000.0     499000  
...                ...        ...  
9499          140000.0     499000  
9500          140000.0     499000  
9501          140000.0     499000  
9502          140000.0     499000  
9503               NaN          0  

[2723 rows x 18 columns]

In [ ]:
#ValueCounts
sales_1['Marketer 2'].value_counts(dropna=False)

NaN              2723
DucVu            1851
YenNguyen02      1748
GiangPhan        1013
NgocNgo           836
TienTran          517
TungTran          512
AnhNguyen08       195
ThuHoang           48
NhiLanh            21
AnhTrinh           11
LinhHoang          10
HuyDo               7
TrangDang           3
LyLuu               3
TuanTran            2
AnhNgo01            1
ht.content.05       1
HuyenNguyen05       1
                    1
Name: Marketer 2, dtype: int64

This column contain Marketers' name, so can't fill in null values and leave them unchanged.

##### **4- Column 'Số lần tương tác'**

In [ ]:
sales_1[sales_1['Số lần tương tác'].isna()]

Ngày       Giờ        Khách hàng          SĐT Channel Chiến dịch  \
204  2020-01-09  14:49:18        Thanh Pham  123400204.0      FB        CVS   
522  2020-10-09  08:38:34        Tran Nhung  123400522.0      FB        CVS   
1865 2020-08-15  15:40:03       Lịch Nguyễn  123401865.0      FB        CVS   
2671 2020-09-17  17:16:32  Lưu Thị Mỹ Huyền  123402671.0      FB        CVS   
3369 2020-02-09  13:46:58           Mai Anh  123403369.0      FB        CVS   
5994 2020-09-28  21:30:09   Nguyễn Mạnh Hải  123405992.0  Upsell     Upsell   
6908 2020-08-30  14:30:12       Yumi Nguyen  123406906.0      FB        CVS   
7068 2020-09-30  15:50:27          Hoan Hao  123407066.0      FB        CVS   

               Content Marketer 2 Type of Lead          Sales  \
204              lop11      DucVu        Tuvan       ThaoKieu   
522   page Việt Phương        NaN        Tuvan    AnhNguyen13   
1865     digheconchong    NgocNgo        Tuvan    AnhNguyen13   
2671       dungpthai-2        NaN        Tuvan         NgocDo   
3369      bosachxizach   TungTran        Tuvan    AnhNguyen13   
5994            Upsell        NaN      Dathang     NganNguyen   
6908             lop11    NgocNgo        Tuvan    HuyenNguyen   
7068     ht.content.05        NaN        Tuvan  HuyenNguyen04   

      Số lần tương tác Trạng thái                               Level  \
204                NaN        NaN                                 NaN   
522                NaN        NaN                                 NaN   
1865               NaN        NaN                                 NaN   
2671               NaN      Trùng                                 NaN   
3369               NaN        NaN                                 NaN   
5994               NaN   Đặt hàng  L8.2 - Khách đã thanh toán toàn bộ   
6908               NaN        NaN                                 NaN   
7068               NaN        NaN                                 NaN   

     Close Date Tỉnh/TP  Số lượng bộ sách  Số tiền giảm giá  Tổng tiền  
204         NaT     NaN               NaN               NaN          0  
522         NaT     NaN               NaN               NaN          0  
1865        NaT     NaN               NaN               NaN          0  
2671        NaT     NaN               NaN               NaN          0  
3369        NaT     NaN               NaN               NaN          0  
5994 2020-09-30     NaN               1.0          140000.0     499000  
6908        NaT     NaN               NaN               NaN          0  
7068        NaT     NaN               NaN               NaN          0

In [ ]:
sales_1['Số lần tương tác'].value_counts(dropna=False)

1.0     7089
2.0      767
3.0      600
4.0      346
5.0      227
6.0      220
7.0      137
8.0       73
9.0       24
10.0       9
NaN        8
11.0       4
Name: Số lần tương tác, dtype: int64

Null values account for a small proportion and can't fill them, so I drop them

In [ ]:
#Xóa giá trị null
sales_1 = sales_1.drop(sales_1[sales_1['Số lần tương tác'].isna()].index)

##### **5- Column 'Type of Lead'**

In [ ]:
sales_1[sales_1['Type of Lead'].isna()]

Ngày       Giờ        Khách hàng          SĐT    Channel  \
124  2020-01-08  15:55:45        Minh Trang  123400124.0         FB   
695  2020-11-07  15:23:21         Chị Tuyến  123400695.0         FB   
1393 2020-08-13  09:12:36  Nguyễn Thu Giang  123401393.0  Saleshunt   
3171 2020-09-19  15:44:19          chị Hiền  123403171.0  Saleshunt   
5437 2020-07-27  09:33:42         Thanh Tâm  123405437.0         FB   
7140 2020-07-31  15:26:54      Phan Mỹ Linh  123407138.0     Upsell   

     Chiến dịch           Content Marketer 2 Type of Lead          Sales  \
124        Mess  leadkhongronguon      DucVu          NaN          AnhHa   
695         CVS           hotline      DucVu          NaN         DatLam   
1393  Saleshunt         Saleshunt  LinhHoang          NaN  HuyenNguyen04   
3171  Saleshunt         Saleshunt        NaN          NaN  HuyenNguyen04   
5437       Mess  leadkhongronguon      DucVu          NaN  HuyenNguyen04   
7140     Upsell            Upsell  LinhHoang          NaN        HoaTran   

      Số lần tương tác Trạng thái                               Level  \
124                1.0   Đặt hàng  L8.2 - Khách đã thanh toán toàn bộ   
695                1.0   Đặt hàng  L8.2 - Khách đã thanh toán toàn bộ   
1393               1.0   Đặt hàng  L8.2 - Khách đã thanh toán toàn bộ   
3171               1.0   Đặt hàng  L8.2 - Khách đã thanh toán toàn bộ   
5437               1.0   Đặt hàng  L8.2 - Khách đã thanh toán toàn bộ   
7140               1.0   Đặt hàng  L8.2 - Khách đã thanh toán toàn bộ   

     Close Date    Tỉnh/TP  Số lượng bộ sách  Số tiền giảm giá  Tổng tiền  
124  2020-08-04     Hà Nội               1.0         -160000.0     799000  
695  2020-07-17   Bắc Ninh               1.0          140000.0     499000  
1393 2020-08-13     Hà Nội               1.0          140000.0     499000  
3171 2020-09-21  Hải Dương               1.0          140000.0     499000  
5437 2020-07-30    Đà Nẵng               1.0          140000.0     499000  
7140 2020-07-28        NaN               1.0          359000.0     280000

In [ ]:
sales_1['Type of Lead'].value_counts(dropna=False)

Dathang               6232
Tuvan                 3257
NaN                      6
Con đang học lớp 7       1
Name: Type of Lead, dtype: int64



This column has 6 null values that account for a small proportion and 1 outlier value, so I drop them

In [ ]:
#Drop Null and Outlier
sales_1 = sales_1.drop(sales_1[(sales_1['Type of Lead'].isna()) | (sales_1['Type of Lead'] == 'Con đang học lớp 7')].index)
sales_1['Type of Lead'].value_counts(dropna=False)

Type of Lead
Dathang    6229
Tuvan      3256
Name: count, dtype: int64

##### **6- Column 'Trạng thái'**

In [ ]:
sales_1[sales_1['Trạng thái'].isna()]

Ngày       Giờ           Khách hàng          SĐT Channel  \
280   2020-10-07  04:29:06              Hai Van  123400282.0      FB   
330   2020-10-07  10:25:33      Nguyễn Thị Thúy  123400332.0      FB   
331   2020-10-07  10:27:06   Nguyễn Thị Anh Đào  123400333.0      FB   
332   2020-10-07  10:33:14    Nguyễn Hồng Nhung  123400334.0      FB   
333   2020-10-07  10:34:27       Nguyen Van Anh  123400335.0      FB   
...          ...       ...                  ...          ...     ...   
9180  2020-09-07  16:35:15         Phạm thị mai  123409193.0      FB   
9181  2020-09-07  16:40:04        Phạm Ngọc Mai  123409194.0      FB   
9182  2020-09-07  16:57:03             Ngọc Để   123409195.0      FB   
9183  2020-09-07  16:58:39  Nguyễn Thị Thu Hiền  123409196.0      FB   
9198  2020-09-07  20:54:23          Đỗ Thị luân  123409211.0      FB   

     Chiến dịch               Content   Marketer 2 Type of Lead  \
280       Brand  docthu_GuidelineBook      NhiLanh      Dathang   
330         CVS            condau_sto  YenNguyen02      Dathang   
331         CVS            condau_sto  YenNguyen02      Dathang   
332         CVS                   bcs  YenNguyen02        Tuvan   
333         CVS            condau_sto  YenNguyen02        Tuvan   
...         ...                   ...          ...          ...   
9180        CVS            condau_sto  YenNguyen02      Dathang   
9181        CVS            condau_sto  YenNguyen02      Dathang   
9182        CVS                  haoi    GiangPhan      Dathang   
9183        CVS            condau_sto  YenNguyen02      Dathang   
9198        CVS           khacbiet-we    GiangPhan      Dathang   

              Sales  Số lần tương tác Trạng thái  \
280          NgocDo               1.0        NaN   
330        ThuyTran               1.0        NaN   
331        ThuyTran               1.0        NaN   
332   HuyenNguyen04               2.0        NaN   
333       HuongDinh               6.0        NaN   
...             ...               ...        ...   
9180       ThuyTran               1.0        NaN   
9181       ThuyTran               1.0        NaN   
9182       ThuyTran               1.0        NaN   
9183       ThuyTran               1.0        NaN   
9198       ThuyTran               1.0        NaN   

                                   Level  Close Date     Tỉnh/TP  \
280   L8.2 - Khách đã thanh toán toàn bộ  2020-07-12      Hà Nội   
330   L8.2 - Khách đã thanh toán toàn bộ  2020-07-13      Hà Nội   
331   L8.2 - Khách đã thanh toán toàn bộ  2020-07-12   Hải Dương   
332                                  NaN         NaN         NaN   
333                                  NaN         NaN         NaN   
...                                  ...         ...         ...   
9180  L8.2 - Khách đã thanh toán toàn bộ  2020-07-13   Khánh Hòa   
9181  L8.2 - Khách đã thanh toán toàn bộ  2020-07-12  Quảng Ninh   
9182  L8.2 - Khách đã thanh toán toàn bộ  2020-07-13  Bình Dương   
9183  L8.2 - Khách đã thanh toán toàn bộ  2020-07-12    Bắc Ninh   
9198  L8.2 - Khách đã thanh toán toàn bộ  2020-07-12      Hà Nội   

      Số lượng bộ sách  Số tiền giảm giá  Tổng tiền  
280                1.0          140000.0     499000  
330                1.0          140000.0     499000  
331                1.0          140000.0     499000  
332                NaN               NaN          0  
333                NaN               NaN          0  
...                ...               ...        ...  
9180               1.0          140000.0     499000  
9181               1.0          140000.0     499000  
9182               1.0          140000.0     499000  
9183               1.0          140000.0     499000  
9198               1.0          140000.0     499000  

[180 rows x 18 columns]

The values in column 'Trạng thái' are quite similar to column 'Level', they both contain customer status information. So I fill in the null values in this column using the information in column 'Level'.

First, the column 'Level' needs to be processed. It can be seen that at the beginning of each value in the column there are confusing characters such as: L8.1, L8.2,... Therefore, these characters need to be removed.

In [ ]:
#Remove characters before '-'
sales_1['Level'] = sales_1['Level'].str.split('-').str[-1]
#Delete the " " character at the beginning
sales_1['Level'] = sales_1['Level'].str.strip()

Check values in column 'Level' when column 'Trạng thái' contains null value

In [ ]:
trangthai_null = sales_1[sales_1['Trạng thái'].isna()]
trangthai_null['Level'].value_counts()

Level
Khách đã thanh toán toàn bộ    136
Hủy đơn sau khi chốt             7
Không nghe máy/Thuê bao          4
Băn khoăn về giá                 1
Trùng                            1
Đã được tư vấn đầy đủ            1
Không còn nhu cầu                1
Name: count, dtype: int64

Column 'Level' contain 7 values, including:

*   Khách đã thanh toán toàn bộ (Lead has paid in full)


*   Hủy đơn sau khi chốt (Cancel order)    
*   Không nghe máy/Thuê bao (Not answer the phone/Unable to contact)

*   Băn khoăn về giá (Confused about the price)

*   Trùng    
*   Đã được tư vấn đầy đủ (Fully consulted)

*   Không còn nhu cầu (no longer need to buy products)   

Next, check when column 'Level' contains these values, what values does column 'Trạng thái' contain






         
         
                 
                       
          




In [ ]:
#FilterLevel
values_to_keep = ['Khách đã thanh toán toàn bộ','Hủy đơn sau khi chốt','Không nghe máy/Thuê bao','Băn khoăn về giá','Trùng','Đã được tư vấn đầy đủ','Không còn nhu cầu']
#FilterData
Filter_data = sales_1[sales_1['Level'].isin(values_to_keep)]
Filter_data
#GroupBy
groupby = Filter_data.groupby(['Level','Trạng thái'])['Trạng thái'].count()
print(groupby)


Level                        Trạng thái               
Băn khoăn về giá             Hẹn lúc khác gọi lại            2
                             Không nghe máy                  5
                             Suy nghĩ thêm/Trả lời sau      31
                             Từ chối                        18
Hủy đơn sau khi chốt         Hủy đơn sau khi chốt          530
                             Suy nghĩ thêm/Trả lời sau       1
                             Từ chối                         3
                             Đặt hàng                      128
Khách đã thanh toán toàn bộ  Đặt hàng                     5952
Không còn nhu cầu            Từ chối                        22
Không nghe máy/Thuê bao      Hẹn lúc khác gọi lại           45
                             Không nghe máy                 92
                             Sai số                          2
                             Suy nghĩ thêm/Trả lời sau      25
                             Thuê bao                       20


Fill in the null values in the column 'Status' with the value that appears most often when observing the value pairs of the columns 'Level' and 'Status'.

For example, when the column 'Level' has the value "Hủy đơn dau khi chốt", I will fill in the 'Status' column with the value "Hủy đơn sau khi chốt" because this pair of values appears the most (530 times). Do the same with the remaining rows.


*   Băn khoăn về giá --> Suy nghĩ thêm/Trả lời sau

*   Hủy đơn sau khi chốt --> Hủy đơn sau khi chốt

*   Khách đã thanh toán toàn bộ --> Đặt hàng
*   Không còn nhu cầu --> Từ chối


*   Không nghe máy/Thuê bao --> Không nghe máy


*   Trùng --> Trùng


*   Đã được tư vấn đầy đủ --> Suy nghĩ thêm/Trả lời sau








  

  



  

  


In [ ]:
#Fill in null
sales_1['Trạng thái'] = sales_1['Trạng thái'].fillna(sales_1['Trạng thái'].where(~(sales_1['Level'] == 'Băn khoăn về giá'), 'Suy nghĩ thêm/Trả lời sau'))
sales_1['Trạng thái'] = sales_1['Trạng thái'].fillna(sales_1['Trạng thái'].where(~(sales_1['Level'] == 'Hủy đơn sau khi chốt'), 'Hủy đơn sau khi chốt'))
sales_1['Trạng thái'] = sales_1['Trạng thái'].fillna(sales_1['Trạng thái'].where(~(sales_1['Level'] == 'Khách đã thanh toán toàn bộ'), 'Đặt hàng'))
sales_1['Trạng thái'] = sales_1['Trạng thái'].fillna(sales_1['Trạng thái'].where(~(sales_1['Level'] == 'Không còn nhu cầu'), 'Từ chối'))
sales_1['Trạng thái'] = sales_1['Trạng thái'].fillna(sales_1['Trạng thái'].where(~(sales_1['Level'] == 'Không nghe máy/Thuê bao'), 'Không nghe máy'))
sales_1['Trạng thái'] = sales_1['Trạng thái'].fillna(sales_1['Trạng thái'].where(~(sales_1['Level'] == 'Trùng'), 'Trùng'))
sales_1['Trạng thái'] = sales_1['Trạng thái'].fillna(sales_1['Trạng thái'].where(~(sales_1['Level'] == 'Đã được tư vấn đầy đủ'), 'Suy nghĩ thêm/Trả lời sau'))

Check for remaining null values

In [ ]:
sales_1['Trạng thái'].isna().sum()

29

Can't fill in these null values, so I drop them

In [ ]:
#Drop
sales_1 = sales_1.drop(sales_1[sales_1['Trạng thái'].isna()].index)


##### **7- Cột 'Level'**





This column contains the same values as column A, so I only need to use one of them for analyzing or reporting. Therefore, drop this column.

In [ ]:
del sales_1['Level']

##### **8- Cột 'Close Date'**

In [ ]:
filter_df = sales_1[sales_1['Close Date'].isna()]
filter_df

Ngày       Giờ           Khách hàng          SĐT Channel  \
0    2020-07-01  00:13:42        Pham thu thuy  123400000.0      FB   
2    2020-07-01  02:16:56       Phạm Thị huong  123400002.0      FB   
3    2020-07-01  05:21:32                  Lan  123400003.0      FB   
6    2020-07-01  05:46:11     Bùi thị hải yến   123400006.0      FB   
9    2020-07-01  06:21:01  Phạm Thị Thúy Hằng   123400009.0      FB   
...         ...       ...                  ...          ...     ...   
9492 2020-09-09  19:54:12          Nguyễn Hằng  123409490.0  Google   
9498 2020-09-09  11:23:30           Trang Than  123409496.0      FB   
9500 2020-09-09  15:30:20      Nguyễn Tiết Hài  123409498.0  Google   
9501 2020-09-09  14:49:21       vương thị hợp   123409499.0  Google   
9503 2020-09-09  20:28:09         Lô Thị Nhung  123409501.0  Google   

     Chiến dịch           Content   Marketer 2 Type of Lead     Sales  \
0           CVS        B1-1812-P1  AnhNguyen08      Dathang  ThuyTran   
2           CVS            matcon      NgocNgo      Dathang  ThuyTran   
3           CVS               cay      NgocNgo      Dathang  ThuyTran   
6           CVS          gockhuat      NgocNgo      Dathang  ThuyTran   
9           CVS          gockhuat      NgocNgo      Dathang  ThuyTran   
...         ...               ...          ...          ...       ...   
9492  Discovery              2010          NaN      Dathang  ThuyTran   
9498       Mess  leadkhongronguon          NaN        Tuvan     AnhHa   
9500     PerMax               909          NaN      Dathang  ThuyTran   
9501     Search              2809          NaN      Dathang  ThuyTran   
9503     Search              2809          NaN        Tuvan   NamPham   

      Số lần tương tác            Trạng thái Close Date Tỉnh/TP  \
0                  1.0  Hủy đơn sau khi chốt        NaT     NaN   
2                  1.0  Hủy đơn sau khi chốt        NaT     NaN   
3                  1.0  Hủy đơn sau khi chốt        NaT     NaN   
6                  1.0  Hủy đơn sau khi chốt        NaT     NaN   
9                  1.0  Hủy đơn sau khi chốt        NaT     NaN   
...                ...                   ...        ...     ...   
9492               1.0              Đặt hàng        NaT     NaN   
9498               3.0  Hẹn lúc khác gọi lại        NaT     NaN   
9500               1.0              Đặt hàng        NaT     NaN   
9501               1.0              Đặt hàng        NaT     NaN   
9503               1.0                Sai số        NaT     NaN   

      Số lượng bộ sách  Số tiền giảm giá  Tổng tiền  
0                  1.0          140000.0     499000  
2                  1.0          140000.0     499000  
3                  1.0          140000.0     499000  
6                  1.0          140000.0     499000  
9                  1.0          140000.0     499000  
...                ...               ...        ...  
9492               1.0          140000.0     499000  
9498               NaN               NaN          0  
9500               1.0          140000.0     499000  
9501               1.0          140000.0     499000  
9503               NaN               NaN          0  

[3308 rows x 17 columns]

This column contains information about the date the customer completes payment, so I think this column contains null values maybe because the payment was not completed. Let's check the Lead status by checking values in column 'Trạng thái'.

In [ ]:
filter_df['Trạng thái'].value_counts()

Trạng thái
Suy nghĩ thêm/Trả lời sau    1119
Hủy đơn sau khi chốt          540
Từ chối                       431
Không nghe máy                421
Đặt hàng                      319
Hẹn lúc khác gọi lại          292
Thuê bao                       81
Sai số                         49
Trùng                          44
Trêu đùa                       12
Name: count, dtype: int64

Column 'Trạng thái' contains mostly values: 'Suy nghĩ thêm/Trả lời sau' (think more/answer later), 'Hủy đơn sau khi chốt' (cancel order), 'Từ chối' (refuse to purchase),...

Therefore, column 'Close date' contains null values when Lead didn't purchase and pay. So I don't drop these null values and leave them unchanged.

##### **9- Column 'Tỉnh/TP'**

In [ ]:
sales_1[sales_1['Tỉnh/TP'].isna()]

Ngày       Giờ           Khách hàng          SĐT Channel  \
0    2020-07-01  00:13:42        Pham thu thuy  123400000.0      FB   
2    2020-07-01  02:16:56       Phạm Thị huong  123400002.0      FB   
3    2020-07-01  05:21:32                  Lan  123400003.0      FB   
6    2020-07-01  05:46:11     Bùi thị hải yến   123400006.0      FB   
9    2020-07-01  06:21:01  Phạm Thị Thúy Hằng   123400009.0      FB   
...         ...       ...                  ...          ...     ...   
9499 2020-09-09  09:28:02      Lê thị thu hiền  123409497.0  Google   
9500 2020-09-09  15:30:20      Nguyễn Tiết Hài  123409498.0  Google   
9501 2020-09-09  14:49:21       vương thị hợp   123409499.0  Google   
9502 2020-09-09  18:44:41         Cao thị hạnh  123409500.0  Google   
9503 2020-09-09  20:28:09         Lô Thị Nhung  123409501.0  Google   

     Chiến dịch              Content   Marketer 2 Type of Lead     Sales  \
0           CVS           B1-1812-P1  AnhNguyen08      Dathang  ThuyTran   
2           CVS               matcon      NgocNgo      Dathang  ThuyTran   
3           CVS                  cay      NgocNgo      Dathang  ThuyTran   
6           CVS             gockhuat      NgocNgo      Dathang  ThuyTran   
9           CVS             gockhuat      NgocNgo      Dathang  ThuyTran   
...         ...                  ...          ...          ...       ...   
9499     PerMax      roasthap_151022          NaN      Dathang  ThuyTran   
9500     PerMax                  909          NaN      Dathang  ThuyTran   
9501     Search                 2809          NaN      Dathang  ThuyTran   
9502     Search  2409-04-01 00:00:00          NaN      Dathang  ThuyTran   
9503     Search                 2809          NaN        Tuvan   NamPham   

      Số lần tương tác            Trạng thái Close Date Tỉnh/TP  \
0                  1.0  Hủy đơn sau khi chốt        NaT     NaN   
2                  1.0  Hủy đơn sau khi chốt        NaT     NaN   
3                  1.0  Hủy đơn sau khi chốt        NaT     NaN   
6                  1.0  Hủy đơn sau khi chốt        NaT     NaN   
9                  1.0  Hủy đơn sau khi chốt        NaT     NaN   
...                ...                   ...        ...     ...   
9499               1.0              Đặt hàng 2020-09-13     NaN   
9500               1.0              Đặt hàng        NaT     NaN   
9501               1.0              Đặt hàng        NaT     NaN   
9502               1.0              Đặt hàng 2020-10-03     NaN   
9503               1.0                Sai số        NaT     NaN   

      Số lượng bộ sách  Số tiền giảm giá  Tổng tiền  
0                  1.0          140000.0     499000  
2                  1.0          140000.0     499000  
3                  1.0          140000.0     499000  
6                  1.0          140000.0     499000  
9                  1.0          140000.0     499000  
...                ...               ...        ...  
9499               1.0          140000.0     499000  
9500               1.0          140000.0     499000  
9501               1.0          140000.0     499000  
9502               1.0          140000.0     499000  
9503               0.0               0.0          0  

[6068 rows x 17 columns]

The column contains information about where the customer lives, so I can't fill in and leave them unchanged.




##### **10- Column 'Số lượng bộ sách' và 'Số tiền giảm giá'**

In [ ]:
#Filter
filter = sales_1[sales_1['Số lượng bộ sách'].isna() & sales_1['Số tiền giảm giá'].isna()]
filter

Ngày       Giờ          Khách hàng          SĐT Channel Chiến dịch  \
15   2020-07-01  08:23:20            Hoàng vy  123400015.0      FB        CVS   
18   2020-07-01  08:28:15            Thanh Hà  123400018.0      FB        CVS   
19   2020-07-01  08:29:58           Ha Nguyen  123400019.0      FB        CVS   
20   2020-07-01  08:31:50             My Hanh  123400020.0      FB        CVS   
26   2020-07-01  09:47:47               Hường  123400026.0      FB        CVS   
...         ...       ...                 ...          ...     ...        ...   
9478 2020-09-09  21:28:55           Ly Nguyen  123409476.0      FB        CVS   
9481 2020-09-09  21:31:52        Huu Anh Dinh  123409479.0      FB        CVS   
9482 2020-09-09  21:36:03  Điện Lạnh Nam Hồng  123409480.0      FB        CVS   
9498 2020-09-09  11:23:30          Trang Than  123409496.0      FB       Mess   
9503 2020-09-09  20:28:09        Lô Thị Nhung  123409501.0  Google     Search   

               Content   Marketer 2 Type of Lead          Sales  \
15            gockhuat      NgocNgo        Tuvan      HuongDinh   
18          B1-1812-P1  AnhNguyen08        Tuvan  HuyenNguyen04   
19          B1-1812-P1  AnhNguyen08        Tuvan      HuongDinh   
20                lop7  AnhNguyen08        Tuvan         NgocDo   
26          condau_sto  YenNguyen02        Tuvan  HuyenNguyen04   
...                ...          ...          ...            ...   
9478             lop11          NaN        Tuvan      SangTrinh   
9481             lop11          NaN        Tuvan        NamPham   
9482           maianh4          NaN        Tuvan      SangTrinh   
9498  leadkhongronguon          NaN        Tuvan          AnhHa   
9503              2809          NaN        Tuvan        NamPham   

      Số lần tương tác                 Trạng thái Close Date Tỉnh/TP  \
15                 6.0  Suy nghĩ thêm/Trả lời sau        NaT     NaN   
18                 2.0  Suy nghĩ thêm/Trả lời sau        NaT     NaN   
19                 7.0             Không nghe máy        NaT     NaN   
20                 2.0             Không nghe máy        NaT     NaN   
26                 2.0  Suy nghĩ thêm/Trả lời sau        NaT     NaN   
...                ...                        ...        ...     ...   
9478               4.0  Suy nghĩ thêm/Trả lời sau        NaT     NaN   
9481               2.0                    Từ chối        NaT     NaN   
9482               4.0       Hẹn lúc khác gọi lại        NaT     NaN   
9498               3.0       Hẹn lúc khác gọi lại        NaT     NaN   
9503               1.0                     Sai số        NaT     NaN   

      Số lượng bộ sách  Số tiền giảm giá  Tổng tiền  
15                 NaN               NaN          0  
18                 NaN               NaN          0  
19                 NaN               NaN          0  
20                 NaN               NaN          0  
26                 NaN               NaN          0  
...                ...               ...        ...  
9478               NaN               NaN          0  
9481               NaN               NaN          0  
9482               NaN               NaN          0  
9498               NaN               NaN          0  
9503               NaN               NaN          0  

[2419 rows x 17 columns]

Like column 'Close Date', column 'Số lượng bộ sách' and 'Số tiền giảm giá contains null values maybe because Lead didn't purchase. Let's check Lead status by checking values in column 'Trạng thái'

In [ ]:
#Value_count
filter['Trạng thái'].value_counts()

Trạng thái
Suy nghĩ thêm/Trả lời sau    1109
Từ chối                       422
Không nghe máy                415
Hẹn lúc khác gọi lại          290
Thuê bao                       80
Sai số                         49
Trùng                          36
Trêu đùa                       12
Đặt hàng                        4
Hủy đơn sau khi chốt            2
Name: count, dtype: int64

The majority of values in this column indicate that the Lead didn't make a purchase. So I fill in the null values in the two columns with values '0'.

In [ ]:
#Fillna
con = (sales_1['Số lượng bộ sách'].isna()) & (sales_1['Số tiền giảm giá'].isna())
sales_1.loc[con, 'Số lượng bộ sách'] = 0
sales_1.loc[con, 'Số tiền giảm giá'] = 0

Check remaining null values in column 'Số tiền giảm giá'

In [ ]:
sales_1['Số tiền giảm giá'].isna().sum()

9

Drop these null values

In [ ]:
#Drop
sales_1 = sales_1.drop(sales_1[sales_1['Số tiền giảm giá'].isna()].index)

### **Change Dtype**

In [ ]:
sales_1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9447 entries, 0 to 9503
Data columns (total 17 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Ngày              9447 non-null   datetime64[ns]
 1   Giờ               9447 non-null   object        
 2   Khách hàng        9447 non-null   object        
 3   SĐT               9447 non-null   float64       
 4   Channel           9447 non-null   object        
 5   Chiến dịch        9447 non-null   object        
 6   Content           9447 non-null   object        
 7   Marketer 2        6740 non-null   object        
 8   Type of Lead      9447 non-null   object        
 9   Sales             9447 non-null   object        
 10  Số lần tương tác  9447 non-null   float64       
 11  Trạng thái        9447 non-null   object        
 12  Close Date        6147 non-null   datetime64[ns]
 13  Tỉnh/TP           3379 non-null   object        
 14  Số lượng bộ sách  9447 non-nu

**Column 'SĐT'**

In [ ]:
sales_1['SĐT'] = sales_1['SĐT'].astype(int)

**Column 'Số lần tương tác'**

In [ ]:
sales_1['Số lần tương tác'] = sales_1['Số lần tương tác'].astype(int)

**Column 'Số lượng bộ sách'**

In [ ]:
sales_1['Số lượng bộ sách'] = sales_1['Số lượng bộ sách'].astype(int)

**Column 'Số tiền giảm giá'**

In [ ]:
sales_1['Số tiền giảm giá'] = sales_1['Số tiền giảm giá'].astype(int)

In [ ]:
#Check Unique Values
sales_1['Số tiền giảm giá'].unique()

array([ 140000,       0,  280000,  240000, -141000, -140000,  128150,
         26173,  139000,  167470,   95850,  107251,   62924,   39186,
       1470000,  389000,  289000,  499000, -233647,  -51000,   10000,
        -10000, -300200, -224841,  420000, -150600, -751000,     970,
        -67801, -259000,   49386, -302400, -214844, -164266, -723130,
        190000,   42987,  152475])

This column contains negative numeric values, this is unreasonable and may be a data entry error, so I convert them to positive values.

In [ ]:
#Change dtype to str
sales_1['Số tiền giảm giá'] = sales_1['Số tiền giảm giá'].astype(str)
#Split the column and get the characters after the '-' sign
sales_1['Số tiền giảm giá'] = sales_1['Số tiền giảm giá'].str.split('-').str[-1]
#Change dtype to int
sales_1['Số tiền giảm giá'] = sales_1['Số tiền giảm giá'].astype(int)

**Check dtype again**

In [ ]:
sales_1.dtypes

Ngày                datetime64[ns]
Giờ                         object
Khách hàng                  object
SĐT                          int64
Channel                     object
Chiến dịch                  object
Content                     object
Marketer 2                  object
Type of Lead                object
Sales                       object
Số lần tương tác             int64
Trạng thái                  object
Close Date          datetime64[ns]
Tỉnh/TP                     object
Số lượng bộ sách             int64
Số tiền giảm giá             int64
Tổng tiền                    int64
dtype: object

In [ ]:
#Save file
sales_1.to_excel('Sales_cleaned.xlsx', index = False)

## **3- Vận đơn (Bill of lading) table**

In [ ]:
#Tạo bản copy
vd_1 = vd.copy()

### **Data information**

In [ ]:
vd_1.shape

(7944, 45)

In [ ]:
vd_1.head()

STT Mã đơn hàng                                   Ghi chú đơn hàng  \
0  1.0    SON25144                                                NaN   
1  NaN         NaN                                                NaN   
2  2.0    SON24883  TẶNG SÁCH LẺ PHẦN 2 CHO PHỤ HUỲNH\ngiao trong ...   
3  NaN         NaN                                                NaN   
4  3.0    SON24943                          Giao trong giờ hành chính   

  Tags đơn hàng Nhân viên tạo đơn           Chi nhánh Nguồn     Mã vận đơn  \
0           NaN  Nguyễn Thị Huyền  Chi nhánh mặc định   Web       G87L96F3   
1           NaN               NaN                 NaN   NaN            NaN   
2           NaN  Đỗ Thị Hồng Ngọc  Chi nhánh mặc định   Web       G87L9VTM   
3           NaN               NaN                 NaN   NaN            NaN   
4           NaN  Đỗ Thị Hồng Ngọc  Chi nhánh mặc định   Web  NSAPOAAAAZPQF   

  Tình trạng gói hàng Trạng thái đối tác  Lý do hủy đơn       Ngày đóng gói  \
0        Đã giao hàng    Giao thành công            NaN 2020-09-16 12:49:54   
1                 NaN                NaN            NaN                 NaT   
2        Đã giao hàng    Giao thành công            NaN 2020-09-16 12:49:13   
3                 NaN                NaN            NaN                 NaT   
4        Đã giao hàng    Giao thành công            NaN 2020-09-15 19:56:20   

   Ngày hẹn giao       Ngày xuất kho      Ngày giao hàng Đối tác giao hàng  \
0            NaN 2020-09-17 11:05:38 2020-09-19 10:20:38      Sapo Express   
1            NaN                 NaT                 NaT               NaN   
2            NaN 2020-09-17 11:05:48 2020-09-21 12:18:59      Sapo Express   
3            NaN                 NaT                 NaT               NaN   
4            NaN 2020-09-15 19:58:37 2020-09-19 12:54:17      Sapo Express   

     Dịch vụ giao hàng  Khối lượng(g)     Kích thước(DxRxC)  Tên người nhận  \
0                Chuẩn          500.0  10.000x10.000x10.000    Mẹ Bống Trần   
1                  NaN            NaN                   NaN             NaN   
2                Chuẩn          500.0  10.000x10.000x10.000         Nhan Le   
3                  NaN            NaN                   NaN             NaN   
4  (Ninja Van - Chuẩn)         1000.0  10.000x10.000x10.000  Vũ Thị Thúy Hà   

  SĐT người nhận                            Địa chỉ giao hàng      Tỉnh/Thành  \
0     0389913230                                    thôn Ngọc        Hưng Yên   
1            NaN                                          NaN             NaN   
2     0907867931         trường mn Phú Thuận đường Phạm Thị B  TP Hồ Chí Minh   
3            NaN                                          NaN             NaN   
4     0916522265  Bệnh viện phụ sản trung ương Triệu Quốc Đạt          Hà Nội   

       Quận/Huyện         Phường xã Trạng thái đối soát  \
0   Huyện Văn Lâm        Xã Lạc Đạo         Đã đối soát   
1             NaN               NaN                 NaN   
2          Quận 7  Phường Phú Thuận         Đã đối soát   
3             NaN               NaN                 NaN   
4  Quận Hoàn Kiếm  Phường Hàng Bông         Đã đối soát   

   Tiền khách phải trả cho đơn  Khách hàng đã trả Hình thức thanh toán  \
0                     499000.0           499000.0                  COD   
1                          NaN                NaN                  NaN   
2                     499000.0           499000.0                  COD   
3                          NaN                NaN                  NaN   
4                     780000.0           780000.0                  COD   

   Tổng tiền thu hộ  Phí vận chuyển Người trả phí  Phí trả đối tác  \
0          499000.0             0.0      Shop trả          26600.0   
1               NaN             NaN           NaN              NaN   
2          499000.0             0.0      Shop trả          26600.0   
3               NaN             NaN           NaN              NaN   
4          780000.0             0.0    

This data has 45 columns, including:

*   STT: serial number column
*   Mã đơn hàng: code orders
*   Ghi chú đơn hàng: order notes
*   Tags đơn hàng: order tags
*   Nhân viên tạo đơn: Name of the employee creating the order
*   Chi nhánh: branch
*   Nguồn: source
*   Mã vận đơn: bill of lading code
*   Tình trạng gói hàng: order status
*   Trạng thái đối tác: shipping partner status
*   Lý do hủy đơn: reason for cancellation
*   Ngày đóng gói: the date the order was packed
*   Ngày hẹn giao hàng: delivery appointment date
*   Ngày xuất kho: date of inventory
*   Ngày giao hàng: delivery date
*   Đối tác giao hàng: name of shipping partner company
*   Dịch vụ giao hàng: name of delivery service
*   Khối lượng: package volume
*   Kích thước(DxRxC): package size
*   Tên người nhận: name of consignee
*   SĐT người nhận: number phone of consignee
*   Địa chỉ giao hàng: delivery address
*   Tỉnh/Thành: the province where the customer lives
*   Quận/Huyện: the district where the customer lives
*   Phường xã: the ward where the customer lives
*   Trạng thái đối soát: Control status
*   Tiền khách phải trả cho đơn: the amount the customer must pay
*   Khách hàng đã trả: the amount the customer has paid
*   Hình thức thanh toán: payments
*   Tổng tiền thu hộ: total amount collected
*   Phí vận chuyển: transport expense
*   Người trả phí: the party must pay the fee
*   Phí trả đối tác: the expense paid to partner company
*   Ghi chú đơn giao: notes for carrier
*   Mã sản phẩm: product code
*   Tên sản phẩm: product name
*   Ghi chú sản phẩm: product note
*   Serial: Unclear
*   Đơn vị tính: unit
*   Đơn giá: unit price
*   CK sản phẩm: product discount
*   CK tổng đơn hàng: order discount
*   Thuế cho từng sản phẩm: tax for each product
*   Tổng tiền hàng: total order amount




In [ ]:
vd_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7944 entries, 0 to 7943
Data columns (total 45 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   STT                          5820 non-null   float64       
 1   Mã đơn hàng                  5820 non-null   object        
 2   Ghi chú đơn hàng             566 non-null    object        
 3   Tags đơn hàng                27 non-null     object        
 4   Nhân viên tạo đơn            5820 non-null   object        
 5   Chi nhánh                    5820 non-null   object        
 6   Nguồn                        5820 non-null   object        
 7   Mã vận đơn                   5820 non-null   object        
 8   Tình trạng gói hàng          5820 non-null   object        
 9   Trạng thái đối tác           5819 non-null   object        
 10  Lý do hủy đơn                0 non-null      float64       
 11  Ngày đóng gói                5820 non-null 

This data has many columns that contain null values.

### **Check duplicate**

In [ ]:
dup_data = vd[vd.duplicated()]
dup_data

STT Mã đơn hàng Ghi chú đơn hàng Tags đơn hàng Nhân viên tạo đơn  \
25    NaN         NaN              NaN           NaN               NaN   
27    NaN         NaN              NaN           NaN               NaN   
95    NaN         NaN              NaN           NaN               NaN   
109   NaN         NaN              NaN           NaN               NaN   
111   NaN         NaN              NaN           NaN               NaN   
...   ...         ...              ...           ...               ...   
7935  NaN         NaN              NaN           NaN               NaN   
7937  NaN         NaN              NaN           NaN               NaN   
7939  NaN         NaN              NaN           NaN               NaN   
7941  NaN         NaN              NaN           NaN               NaN   
7943  NaN         NaN              NaN           NaN               NaN   

     Chi nhánh Nguồn Mã vận đơn Tình trạng gói hàng Trạng thái đối tác  \
25         NaN   NaN        NaN                 NaN                NaN   
27         NaN   NaN        NaN                 NaN                NaN   
95         NaN   NaN        NaN                 NaN                NaN   
109        NaN   NaN        NaN                 NaN                NaN   
111        NaN   NaN        NaN                 NaN                NaN   
...        ...   ...        ...                 ...                ...   
7935       NaN   NaN        NaN                 NaN                NaN   
7937       NaN   NaN        NaN                 NaN                NaN   
7939       NaN   NaN        NaN                 NaN                NaN   
7941       NaN   NaN        NaN                 NaN                NaN   
7943       NaN   NaN        NaN                 NaN                NaN   

      Lý do hủy đơn Ngày đóng gói  Ngày hẹn giao Ngày xuất kho Ngày giao hàng  \
25              NaN           NaT            NaN           NaT            NaT   
27              NaN           NaT            NaN           NaT            NaT   
95              NaN           NaT            NaN           NaT            NaT   
109             NaN           NaT            NaN           NaT            NaT   
111             NaN           NaT            NaN           NaT            NaT   
...             ...           ...            ...           ...            ...   
7935            NaN           NaT            NaN           NaT            NaT   
7937            NaN           NaT            NaN           NaT            NaT   
7939            NaN           NaT            NaN           NaT            NaT   
7941            NaN           NaT            NaN           NaT            NaT   
7943            NaN           NaT            NaN           NaT            NaT   

     Đối tác giao hàng Dịch vụ giao hàng  Khối lượng(g) Kích thước(DxRxC)  \
25                 NaN               NaN            NaN               NaN   
27                 NaN               NaN            NaN               NaN   
95                 NaN               NaN            NaN               NaN   
109                NaN               NaN            NaN               NaN   
111                NaN               NaN            NaN               NaN   
...                ...               ...            ...               ...   
7935               NaN               NaN            NaN               NaN   
7937               NaN               NaN            NaN               NaN   
7939               NaN               NaN            NaN               NaN   
7941               NaN               NaN            NaN               NaN   
7943               NaN               NaN            NaN               NaN   

     Tên người nhận SĐT người nhận Địa chỉ giao hàng Tỉnh/Thành Quận/Huyện  \
25              NaN            NaN               NaN        NaN        NaN   
27              NaN            NaN               NaN        NaN        NaN   
95              NaN            NaN               NaN        NaN        NaN   
109             Na

In [ ]:
#Drop duplicate
vd_1 = vd_1.drop_duplicates()

### **Handle null values**

In [ ]:
#Drop column 'STT'
vd_1.drop(columns = ['STT'], inplace = True)

In [ ]:
vd_1.isna().sum()

Mã đơn hàng                      24
Ghi chú đơn hàng               5278
Tags đơn hàng                  5817
Nhân viên tạo đơn                24
Chi nhánh                        24
Nguồn                            24
Mã vận đơn                       24
Tình trạng gói hàng              24
Trạng thái đối tác               25
Lý do hủy đơn                  5844
Ngày đóng gói                    24
Ngày hẹn giao                  5844
Ngày xuất kho                    24
Ngày giao hàng                  729
Đối tác giao hàng                24
Dịch vụ giao hàng                25
Khối lượng(g)                    24
Kích thước(DxRxC)                24
Tên người nhận                   24
SĐT người nhận                   24
Địa chỉ giao hàng                24
Tỉnh/Thành                       24
Quận/Huyện                       24
Phường xã                        24
Trạng thái đối soát              24
Tiền khách phải trả cho đơn      24
Khách hàng đã trả               811
Hình thức thanh toán        



Many columns have 24 null values, so let's check this data area.

In [ ]:
df = vd_1[vd_1['Mã đơn hàng'].isna()]
df

Mã đơn hàng Ghi chú đơn hàng Tags đơn hàng Nhân viên tạo đơn Chi nhánh  \
1            NaN              NaN           NaN               NaN       NaN   
3            NaN              NaN           NaN               NaN       NaN   
15           NaN              NaN           NaN               NaN       NaN   
16           NaN              NaN           NaN               NaN       NaN   
23           NaN              NaN           NaN               NaN       NaN   
42           NaN              NaN           NaN               NaN       NaN   
101          NaN              NaN           NaN               NaN       NaN   
215          NaN              NaN           NaN               NaN       NaN   
425          NaN              NaN           NaN               NaN       NaN   
442          NaN              NaN           NaN               NaN       NaN   
2285         NaN              NaN           NaN               NaN       NaN   
2700         NaN              NaN           NaN               NaN       NaN   
3510         NaN              NaN           NaN               NaN       NaN   
4875         NaN              NaN           NaN               NaN       NaN   
4887         NaN              NaN           NaN               NaN       NaN   
5003         NaN              NaN           NaN               NaN       NaN   
5141         NaN              NaN           NaN               NaN       NaN   
5241         NaN              NaN           NaN               NaN       NaN   
5685         NaN              NaN           NaN               NaN       NaN   
5686         NaN              NaN           NaN               NaN       NaN   
6247         NaN              NaN           NaN               NaN       NaN   
6769         NaN              NaN           NaN               NaN       NaN   
6904         NaN              NaN           NaN               NaN       NaN   
7086         NaN              NaN           NaN               NaN       NaN   

     Nguồn Mã vận đơn Tình trạng gói hàng Trạng thái đối tác  Lý do hủy đơn  \
1      NaN        NaN                 NaN                NaN            NaN   
3      NaN        NaN                 NaN                NaN            NaN   
15     NaN        NaN                 NaN                NaN            NaN   
16     NaN        NaN                 NaN                NaN            NaN   
23     NaN        NaN                 NaN                NaN            NaN   
42     NaN        NaN                 NaN                NaN            NaN   
101    NaN        NaN                 NaN                NaN            NaN   
215    NaN        NaN                 NaN                NaN            NaN   
425    NaN        NaN                 NaN                NaN            NaN   
442    NaN        NaN                 NaN                NaN            NaN   
2285   NaN        NaN                 NaN                NaN            NaN   
2700   NaN        NaN                 NaN                NaN            NaN   
3510   NaN        NaN                 NaN                NaN            NaN   
4875   NaN        NaN                 NaN                NaN            NaN   
4887   NaN        NaN                 NaN                NaN            NaN   
5003   NaN        NaN                 NaN                NaN            NaN   
5141   NaN        NaN                 NaN                NaN            NaN   
5241   NaN        NaN                 NaN                NaN            NaN   
5685   NaN        NaN                 NaN                NaN            NaN   
5686   NaN        NaN                 NaN                NaN            NaN   
6247   NaN        NaN                 NaN                NaN            NaN   
6769   NaN        NaN                 NaN                NaN            NaN   
6904   NaN        NaN                 NaN                NaN            NaN   
7086   NaN        NaN                 NaN                NaN            NaN   

     Ngày đóng gói  Ngày hẹn giao Ngày xuất kho Ngày 

In [ ]:
#null rate of columns
df.isna().mean()

Mã đơn hàng                    1.000000
Ghi chú đơn hàng               1.000000
Tags đơn hàng                  1.000000
Nhân viên tạo đơn              1.000000
Chi nhánh                      1.000000
Nguồn                          1.000000
Mã vận đơn                     1.000000
Tình trạng gói hàng            1.000000
Trạng thái đối tác             1.000000
Lý do hủy đơn                  1.000000
Ngày đóng gói                  1.000000
Ngày hẹn giao                  1.000000
Ngày xuất kho                  1.000000
Ngày giao hàng                 1.000000
Đối tác giao hàng              1.000000
Dịch vụ giao hàng              1.000000
Khối lượng(g)                  1.000000
Kích thước(DxRxC)              1.000000
Tên người nhận                 1.000000
SĐT người nhận                 1.000000
Địa chỉ giao hàng              1.000000
Tỉnh/Thành                     1.000000
Quận/Huyện                     1.000000
Phường xã                      1.000000
Trạng thái đối soát            1.000000


This data area has up to 35 out of 44 columns containing 100 percent null values. So I drop it from the dataset.

In [ ]:
vd_1 = vd_1.drop(vd_1[vd_1['Mã đơn hàng'].isna()].index)

In [ ]:
#Check shape
vd_1.shape

(5820, 44)

In [ ]:
#Check null values again
vd_1.isna().sum()

Mã đơn hàng                       0
Ghi chú đơn hàng               5254
Tags đơn hàng                  5793
Nhân viên tạo đơn                 0
Chi nhánh                         0
Nguồn                             0
Mã vận đơn                        0
Tình trạng gói hàng               0
Trạng thái đối tác                1
Lý do hủy đơn                  5820
Ngày đóng gói                     0
Ngày hẹn giao                  5820
Ngày xuất kho                     0
Ngày giao hàng                  705
Đối tác giao hàng                 0
Dịch vụ giao hàng                 1
Khối lượng(g)                     0
Kích thước(DxRxC)                 0
Tên người nhận                    0
SĐT người nhận                    0
Địa chỉ giao hàng                 0
Tỉnh/Thành                        0
Quận/Huyện                        0
Phường xã                         0
Trạng thái đối soát               0
Tiền khách phải trả cho đơn       0
Khách hàng đã trả               787
Hình thức thanh toán        

**Drop some columns that have almost or all null values:** 'Ghi chú đơn hàng', 'Tags đơn hàng', 'Lý do hủy đơn', 'Ngày hẹn giao', 'Ghi chú sản phẩm', 'Serial'.

In [ ]:
#Drop Columns
columns_to_drop = ['Ghi chú đơn hàng','Tags đơn hàng','Lý do hủy đơn','Ngày hẹn giao','Ghi chú sản phẩm','Serial','Đơn vị tính']
vd_1.drop(columns = columns_to_drop, inplace = True)

In [ ]:
#Check null values again
vd_1.isna().sum()

Mã đơn hàng                      0
Nhân viên tạo đơn                0
Chi nhánh                        0
Nguồn                            0
Mã vận đơn                       0
Tình trạng gói hàng              0
Trạng thái đối tác               1
Ngày đóng gói                    0
Ngày xuất kho                    0
Ngày giao hàng                 705
Đối tác giao hàng                0
Dịch vụ giao hàng                1
Khối lượng(g)                    0
Kích thước(DxRxC)                0
Tên người nhận                   0
SĐT người nhận                   0
Địa chỉ giao hàng                0
Tỉnh/Thành                       0
Quận/Huyện                       0
Phường xã                        0
Trạng thái đối soát              0
Tiền khách phải trả cho đơn      0
Khách hàng đã trả              787
Hình thức thanh toán           787
Tổng tiền thu hộ                 0
Phí vận chuyển                   0
Người trả phí                    0
Phí trả đối tác                  0
Ghi chú đơn giao    

#### **Check each column that contain null values**

**1- Column 'Trạng thái đối tác' and 'Dịch vụ giao hàng'**

In [ ]:
vd_1[vd_1['Trạng thái đối tác'].isna()]

Mã đơn hàng Nhân viên tạo đơn           Chi nhánh Nguồn Mã vận đơn  \
3036    SON23116    Đinh Thu Hường  Chi nhánh mặc định   Web   FUN25838   

     Tình trạng gói hàng Trạng thái đối tác       Ngày đóng gói  \
3036        Đã giao hàng                NaN 2020-08-25 15:24:21   

           Ngày xuất kho      Ngày giao hàng Đối tác giao hàng  \
3036 2020-08-25 15:24:24 2020-09-20 18:48:55        SHIP NGOÀI   

     Dịch vụ giao hàng  Khối lượng(g)     Kích thước(DxRxC)  \
3036               NaN          500.0  10.000x10.000x10.000   

         Tên người nhận SĐT người nhận  \
3036  Chị Vũ Hồng Nhung     0388252938   

                                    Địa chỉ giao hàng       Tỉnh/Thành  \
3036  số 80/19 đường Nguyễn Mạnh Hùng, tổ 7 khu phố 1  Bà Rịa-Vũng Tàu   

            Quận/Huyện         Phường xã Trạng thái đối soát  \
3036  Thành phố Bà Rịa  Phường Long Toàn       Chưa đối soát   

      Tiền khách phải trả cho đơn  Khách hàng đã trả Hình thức thanh toán  \
3036                     499000.0           499000.0                  COD   

      Tổng tiền thu hộ  Phí vận chuyển Người trả phí  Phí trả đối tác  \
3036          499000.0             0.0      Shop trả              0.0   

                                       Ghi chú đơn giao Mã sản phẩm  \
3036  Cho xem hàng, cho thử |đã tạo đơn bên nhanh.vn...        php3   

                   Tên sản phẩm  Số lượng  Đơn giá  CK sản phẩm  \
3036  Sách TeenUp - PH - Phần 3         1        0            0   

      CK tổng đơn hàng  Thuế cho từng sản phẩm  Tổng tiền hàng  
3036                 0                       0               0

This column has 1 null values. Drop this row.

In [ ]:
df = vd_1[vd_1['Dịch vụ giao hàng'].isna()]
df

Mã đơn hàng Nhân viên tạo đơn           Chi nhánh Nguồn Mã vận đơn  \
3036    SON23116    Đinh Thu Hường  Chi nhánh mặc định   Web   FUN25838   

     Tình trạng gói hàng Trạng thái đối tác       Ngày đóng gói  \
3036        Đã giao hàng                NaN 2020-08-25 15:24:21   

           Ngày xuất kho      Ngày giao hàng  ... Phí trả đối tác  \
3036 2020-08-25 15:24:24 2020-09-20 18:48:55  ...             0.0   

                                       Ghi chú đơn giao  Mã sản phẩm  \
3036  Cho xem hàng, cho thử |đã tạo đơn bên nhanh.vn...         php3   

                   Tên sản phẩm Số lượng Đơn giá CK sản phẩm CK tổng đơn hàng  \
3036  Sách TeenUp - PH - Phần 3        1       0           0                0   

     Thuế cho từng sản phẩm Tổng tiền hàng  
3036                      0              0  

[1 rows x 37 columns]

In [ ]:
vd_1 = vd_1.drop(vd_1[vd_1['Trạng thái đối tác'].isna()].index)

In [ ]:
df = vd_1[vd_1['Dịch vụ giao hàng'].isna()]
df

Mã đơn hàng Nhân viên tạo đơn           Chi nhánh Nguồn Mã vận đơn  \
3036    SON23116    Đinh Thu Hường  Chi nhánh mặc định   Web   FUN25838   

     Tình trạng gói hàng Trạng thái đối tác       Ngày đóng gói  \
3036        Đã giao hàng                NaN 2020-08-25 15:24:21   

           Ngày xuất kho      Ngày giao hàng  ... Phí trả đối tác  \
3036 2020-08-25 15:24:24 2020-09-20 18:48:55  ...             0.0   

                                       Ghi chú đơn giao  Mã sản phẩm  \
3036  Cho xem hàng, cho thử |đã tạo đơn bên nhanh.vn...         php3   

                   Tên sản phẩm Số lượng Đơn giá CK sản phẩm CK tổng đơn hàng  \
3036  Sách TeenUp - PH - Phần 3        1       0           0                0   

     Thuế cho từng sản phẩm Tổng tiền hàng  
3036                      0              0  

[1 rows x 37 columns]

This column has 1 null values. Drop this row.

In [ ]:
vd_1 = vd_1.drop(vd_1[vd_1['Dịch vụ giao hàng'].isna()].index)

**2- Column 'Ngày giao hàng'**

In [ ]:
df = vd_1[vd_1['Ngày giao hàng'].isna()]
df.head(5)

Mã đơn hàng Nhân viên tạo đơn           Chi nhánh Nguồn     Mã vận đơn  \
11    SON25391     Trần Thị Thúy  Chi nhánh mặc định   Web  NSAPOAAAAZPPX   
12    SON25304     Trần Thị Thúy  Chi nhánh mặc định   Web  NSAPOAAAAZPPW   
21    SON25749     Trần Thị Thúy  Chi nhánh mặc định   Web       G87Q4FUC   
47    SON25791     Trần Thị Thúy  Chi nhánh mặc định   Web       G87QBURA   
75    SON25833     Trần Thị Thúy  Chi nhánh mặc định   Web       G87QBG7W   

   Tình trạng gói hàng Trạng thái đối tác       Ngày đóng gói  \
11    Hủy giao-đã nhận     Trả thành công 2020-09-15 19:08:50   
12    Hủy giao-đã nhận        Chờ lấy lại 2020-09-15 19:08:35   
21    Hủy giao-đã nhận     Trả thành công 2020-09-15 17:30:05   
47    Hủy giao-đã nhận     Trả thành công 2020-09-15 16:37:40   
75    Hủy giao-đã nhận     Trả thành công 2020-09-15 16:36:05   

         Ngày xuất kho Ngày giao hàng Đối tác giao hàng  \
11 2020-09-15 19:39:08            NaT      Sapo Express   
12 2020-09-15 19:39:28            NaT      Sapo Express   
21 2020-09-15 17:34:49            NaT      Sapo Express   
47 2020-09-15 17:03:32            NaT      Sapo Express   
75 2020-09-15 17:07:52            NaT      Sapo Express   

        Dịch vụ giao hàng  Khối lượng(g)     Kích thước(DxRxC)  \
11                  Chuẩn          500.0  10.000x10.000x10.000   
12                  Chuẩn          500.0  10.000x10.000x10.000   
21  (GHN Express - Chuẩn)          500.0  10.000x10.000x10.000   
47  (GHN Express - Chuẩn)          500.0  10.000x10.000x10.000   
75  (GHN Express - Chuẩn)          500.0  10.000x10.000x10.000   

     Tên người nhận SĐT người nhận  \
11  Nguyễn thị thúy     0345269989   
12  Nguyễn thị thúy     0345269989   
21    Phung duc nam     0918736749   
47     Nguyễn ly na     0915992835   
75     Trang Nguyen     0983235102   

                                    Địa chỉ giao hàng      Tỉnh/Thành  \
11  Phường Thọ Quang Quận Sơn Trà Đà Nẵng Việt Nam...         Đà Nẵng   
12  Phường Thọ Quang Quận Sơn Trà Đà Nẵng Việt Nam...         Đà Nẵng   
21  Phường Võ Thị Sáu Quận 3 Hồ Chí Minh Việt Nam ...  TP Hồ Chí Minh   
47  Phường Nam Lý Thành phố Đồng Hới Quảng Bình Vi...      Quảng Bình   
75  Phường Hoàng Văn Thụ Quận Hoàng Mai Hà Nội Việ...          Hà Nội   

            Quận/Huyện             Phường xã Trạng thái đối soát  \
11        Quận Sơn Trà      Phường Thọ Quang         Đã đối soát   
12        Quận Sơn Trà      Phường Thọ Quang       Chưa đối soát   
21              Quận 3     Phường Võ Thị Sáu         Đã đối soát   
47  Thành Phố Đồng Hới         Phường Nam Lý         Đã đối soát   
75      Quận Hoàng Mai  Phường Hoàng Văn Thụ         Đã đối soát   

    Tiền khách phải trả cho đơn  Khách hàng đã trả Hình thức thanh toán  \
11                     499000.0                NaN                  NaN   
12                     499000.0                NaN                  NaN   
21                     499000.0                NaN                  NaN   
47                     499000.0                NaN                  NaN   
75                     499000.0                NaN                  NaN   

    Tổng tiền thu hộ  Phí vận chuyển Người trả phí  Phí trả đối tác  \
11          499000.0             0.0      Shop trả          22000.0   
12          499000.0             0.0      Shop trả          20800.0   
21          499000.0             0.0      Shop trả          26600.0   
47          499000.0             0.0      Shop trả          26600.0   
75          499000.0             0.0      Shop trả          18955.0   

   Ghi chú đơn giao     Mã sản phẩm          Tên sản phẩm  Số lượng  Đơn giá  \
11     Cho thử hàng  SACHTEENUPKID2  BỘ SÁCH TEENUP - KID         1   639000   
12     Cho thử hàng  SACHTEENUPKID2  BỘ SÁCH TEENUP - KID         1   639000   
21     Cho thử hàng  SACHTEENUPKID2  BỘ SÁCH TEENUP - KID         1   639000   
47     Cho thử hàng  SACHTEENUPKID2  BỘ SÁCH TEENUP - KID         1   639000   
75     Cho thử hàng  SACHTEENUPKID2  BỘ SÁ

This column has 705 null values, the reason may be because the customer didn't receive the order. Let's check order status

In [ ]:
df['Tình trạng gói hàng'].unique()

array(['Hủy giao-đã nhận', 'Đang giao hàng', 'Chờ giao lại'], dtype=object)

When column 'Ngày giao hàng' contains null values, order status is one of the following values: 'Hủy giao-đã nhận' (cancel delivery), 'Đang giao hàng' (delivering), 'Chờ giao lại' (waiting for delivery). It can be concluded that column 'Ngày giao hàng' contains null values because the customer refused or didn't receive the order. So don't need to drop these null values and leave them unchanged.

**3- Columns 'Khách hàng đã trả' and 'Hình thức thanh toán'**

In [ ]:
df = vd_1[vd_1['Khách hàng đã trả'].isna() & vd_1['Hình thức thanh toán'].isna()]
df

Mã đơn hàng    Nhân viên tạo đơn           Chi nhánh Nguồn  \
10      SON25590  Đinh Thục Hoàng Anh  Chi nhánh mặc định   Web   
11      SON25391        Trần Thị Thúy  Chi nhánh mặc định   Web   
12      SON25304        Trần Thị Thúy  Chi nhánh mặc định   Web   
13      SON25136        Đỗ Khánh Linh  Chi nhánh mặc định   Web   
14      SON24843   Trương Minh Nguyệt  Chi nhánh mặc định   Web   
...          ...                  ...                 ...   ...   
7924    SON20019        Trần Thị Thúy  Chi nhánh mặc định   Web   
7926    SON20027        Trần Thị Thúy  Chi nhánh mặc định   Web   
7930    SON20034        Trần Thị Thúy  Chi nhánh mặc định   Web   
7934    SON20036  Đinh Thục Hoàng Anh  Chi nhánh mặc định   Web   
7940    SON20040       Đinh Thu Hường  Chi nhánh mặc định   Web   

         Mã vận đơn Tình trạng gói hàng Trạng thái đối tác  \
10         G87QKEGR        Đã giao hàng    Giao thành công   
11    NSAPOAAAAZPPX    Hủy giao-đã nhận     Trả thành công   
12    NSAPOAAAAZPPW    Hủy giao-đã nhận        Chờ lấy lại   
13         G87QKM8B        Đã giao hàng    Giao thành công   
14         G87QKM3Q        Đã giao hàng    Giao thành công   
...             ...                 ...                ...   
7924       G8QL49HC    Hủy giao-đã nhận     Trả thành công   
7926       G8QL49TC    Hủy giao-đã nhận     Trả thành công   
7930       G8QL49KL    Hủy giao-đã nhận     Trả thành công   
7934       G8QL493P        Đã giao hàng    Giao thành công   
7940       G8QL49NW    Hủy giao-đã nhận     Trả thành công   

           Ngày đóng gói       Ngày xuất kho      Ngày giao hàng  \
10   2020-09-15 19:09:06 2020-09-15 19:42:26 2020-09-19 14:12:25   
11   2020-09-15 19:08:50 2020-09-15 19:39:08                 NaT   
12   2020-09-15 19:08:35 2020-09-15 19:39:28                 NaT   
13   2020-09-15 19:08:12 2020-09-15 19:41:49 2020-09-18 20:12:15   
14   2020-09-15 19:07:00 2020-09-15 19:41:17 2020-09-17 11:37:03   
...                  ...                 ...                 ...   
7924 2020-07-03 12:12:21 2020-07-03 15:43:56                 NaT   
7926 2020-07-03 12:12:20 2020-07-03 15:44:08                 NaT   
7930 2020-07-03 12:12:18 2020-07-03 15:44:37                 NaT   
7934 2020-07-03 12:12:16 2020-07-03 15:45:16 2020-07-05 09:32:45   
7940 2020-07-03 12:12:13 2020-07-03 15:46:15                 NaT   

     Đối tác giao hàng      Dịch vụ giao hàng  Khối lượng(g)  \
10        Sapo Express                  Chuẩn          100.0   
11        Sapo Express                  Chuẩn          500.0   
12        Sapo Express                  Chuẩn          500.0   
13        Sapo Express                  Chuẩn          100.0   
14        Sapo Express                  Chuẩn          100.0   
...                ...                    ...            ...   
7924      Sapo Express  (GHN Express - Chuẩn)          500.0   
7926      Sapo Express  (GHN Express - Chuẩn)          500.0   
7930      Sapo Express  (GHN Express - Chuẩn)          500.0   
7934      Sapo Express  (GHN Express - Chuẩn)          500.0   
7940      Sapo Express  (GHN Express - Chuẩn)          500.0   

         Kích thước(DxRxC)   Tên người nhận SĐT người nhận  \
10    10.000x10.000x10.000         Chị Dung     0832353580   
11    10.000x10.000x10.000  Nguyễn thị thúy     0345269989   
12    10.000x10.000x10.000  Nguyễn thị thúy     0345269989   
13    10.000x10.000x10.000      Hoàng Giang     0363843461   
14    10.000x10.000x10.000     Mẹ Thúy Phạm     0769291102   
...                    ...              ...            ...   
7924  10.000x10.000x10.000     đặng thị hải     0325636163   
7926  10.000x10.000x10.000     Trần thị thư     0989707212   
7930  10.000x10.000x10.000    Pham thu thuy     0903080381   
7934  10.000x10.000x10.000  Nguyễn Thị Hoàn     0967853910   
7940  10.000x10.000x10.000  Chị Lê Minh Tâm     0366732179   

                                      Địa chỉ giao hàng      Tỉnh/Thành  \
10    Phường An Lợi Đông Thành Phố Thủ Đức Hồ Chí

Like column 'Ngày giao hàng', these two columns contain null values because customers refused or didn't receive the order. Leave these null values unchanged.

In [ ]:
#Save file
vd_1.to_excel('Vận đơn_cleaned.xlsx',index = False)

# **Create some additional columns for analyzing and reporting**

**1- Thời gian giao hàng**

Located in the 'Vận đơn' table, contain information about the delivery time, from the time the order is shipped to the time the customer receives the order.

In [ ]:
#Add column
vd_1['Thời gian giao hàng (giờ)'] = vd_1['Ngày giao hàng'] - vd_1['Ngày xuất kho']

In [ ]:
vd_1['Thời gian giao hàng (giờ)']

0      1 days 23:15:00
1      4 days 01:13:11
2      3 days 16:55:40
3      2 days 18:47:48
4      2 days 15:04:33
             ...      
5814   1 days 17:47:29
5815   2 days 00:06:36
5816   0 days 20:45:57
5817               NaT
5818   0 days 18:01:57
Name: Thời gian giao hàng, Length: 5819, dtype: timedelta64[ns]

In [ ]:
#Convert the value in the column to hour
from datetime import timedelta
vd_1['Thời gian giao hàng (giờ)'] = vd_1['Thời gian giao hàng (giờ)'].dt.total_seconds() / 3600
vd_1['Thời gian giao hàng (giờ)']

0       47.250000
1       97.219722
2       88.927778
3       66.796667
4       63.075833
          ...    
5814    41.791389
5815    48.110000
5816    20.765833
5817          NaN
5818    18.032500
Name: Thời gian giao hàng, Length: 5819, dtype: float64

In [ ]:
vd_1['Thời gian giao hàng (giờ)'] = round(vd_1['Thời gian giao hàng (giờ)'], 0)
vd_1['Thời gian giao hàng (giờ)']

0       47.0
1       97.0
2       89.0
3       67.0
4       63.0
        ... 
5814    42.0
5815    48.0
5816    21.0
5817     NaN
5818    18.0
Name: Thời gian giao hàng, Length: 5819, dtype: float64

**2- Miền**

Delivery location by region: North, Central, South

In [ ]:
domain_mapping = {
    'Miền Bắc': ['Phú Thọ','Hải Dương','Bắc Ninh','Bắc Giang','Cao Bằng','Điện Biên','Hải Phòng','Sơn La','Hòa Bình','Vĩnh Phúc','Hà Nội','Lạng Sơn','Bắc Kạn','Lào Cai','Quảng Ninh','Hà Nam','Hà Giang','Tuyên Quang','Lai Châu','Thái Nguyên','Ninh Bình','Nam Định','Hưng Yên','Yên Bái','Thái Bình'],
    'Miền Trung': ['Quảng Ngãi','Lâm Đồng','Khánh Hòa','Thanh Hóa','Đà Nẵng','Đắk Nông','Quảng Nam','Nghệ An','Quảng Bình','Bình Định','Kon Tum','Phú Yên','Đắk Lắk','Quảng Trị','Hà Tĩnh','Gia Lai','Thừa Thiên Huế'],
    'Miền Nam': ['Trà Vinh','Bình Thuận','Hậu Giang','Bạc Liêu','Ninh Thuận','Vĩnh Long','An Giang','Đồng Nai','Tây Ninh','Tiền Giang','Bà Rịa-Vũng Tàu','Kiên Giang','Bến Tre','Bình Dương','TP Hồ Chí Minh','Bình Phước','Đồng Tháp','Cần Thơ','Long An','Cà Mau','Sóc Trăng']
}
vd_1['Miền'] = vd_1['Tỉnh/Thành'].apply(lambda x: next((key for key, value in domain_mapping.items() if x in value), None))

In [ ]:
vd_1[['Miền','Tỉnh/Thành']]

Miền      Tỉnh/Thành
0       Miền Bắc        Hưng Yên
1       Miền Nam  TP Hồ Chí Minh
2       Miền Bắc          Hà Nội
3     Miền Trung       Thanh Hóa
4     Miền Trung       Quảng Nam
...          ...             ...
5814  Miền Trung         Nghệ An
5815    Miền Nam  TP Hồ Chí Minh
5816    Miền Bắc       Hải Phòng
5817    Miền Bắc        Nam Định
5818    Miền Bắc         Phú Thọ

[5819 rows x 2 columns]